File: collect_app/src/main/java/org/odk/collect/android/preferences/screens/MapsPreferencesFragment.java
/*
 * Copyright (C) 2017 Shobhit
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package org.odk.collect.android.preferences.screens;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.DialogFragment;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceCategory;

import org.odk.collect.android.R;
import org.odk.collect.android.geo.MapConfigurator;
import org.odk.collect.android.geo.MapProvider;
import org.odk.collect.android.preferences.CaptionedListPreference;
import org.odk.collect.android.preferences.CaptionedListPreference.Item;
import org.odk.collect.android.preferences.PrefUtils;
import org.odk.collect.android.preferences.dialogs.ReferenceLayerPreferenceDialog;
import org.odk.collect.android.storage.StoragePathProvider;
import org.odk.collect.android.storage.StorageSubdirectory;
import org.odk.collect.android.utilities.FileUtils;
import org.odk.collect.android.utilities.MultiClickGuard;
import org.odk.collect.shared.PathUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.odk.collect.android.preferences.keys.ProjectKeys.CATEGORY_BASEMAP;
import static org.odk.collect.android.preferences.keys.ProjectKeys.KEY_BASEMAP_SOURCE;

public class MapsPreferencesFragment extends BaseProjectPreferencesFragment {

    private Context context;
    private ListPreference basemapSourcePref;
    private CaptionedListPreference referenceLayerPref;
    private boolean autoShowReferenceLayerDialog;

    @Override
    public void onDisplayPreferenceDialog(Preference preference) {
        if (MultiClickGuard.allowClick(getClass().getName())) {
            DialogFragment dialogFragment = null;
            if (preference instanceof CaptionedListPreference) {
                dialogFragment = ReferenceLayerPreferenceDialog.newInstance(preference.getKey());
            } else {
                super.onDisplayPreferenceDialog(preference);
            }
            if (dialogFragment != null) {
                dialogFragment.setTargetFragment(this, 0);
                dialogFragment.show(getParentFragmentManager(), ReferenceLayerPreferenceDialog.class.getName());
            }
        }
    }

    /** Pops up the preference dialog that lets the user choose a reference layer. */
    public static void showReferenceLayerDialog(Activity activity) {
        // Unfortunately, the Preference class is designed so that it is impossible
        // to just open a preference dialog without building a PreferenceFragment
        // and attaching it to an activity.  So, we instantiate a MapsPreference
        // fragment that is configured to immediately open the dialog when it's
        // attached, then instantiate it and attach it.
        MapsPreferencesFragment prefs = new MapsPreferencesFragment();
        prefs.autoShowReferenceLayerDialog = true;  // makes dialog open immediately
        ((AppCompatActivity) activity).getSupportFragmentManager()
            .beginTransaction()
            .add(prefs, null)
            .commit();
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        super.onCreatePreferences(savedInstanceState, rootKey);
        setPreferencesFromResource(R.xml.maps_preferences, rootKey);

        context = getPreferenceScreen().getContext();
        initBasemapSourcePref();
        initReferenceLayerPref();
        if (autoShowReferenceLayerDialog) {
            populateReferenceLayerPref();
            /** Opens the dialog programmatically, rather than by a click from the user. */
            onDisplayPreferenceDialog(getPreferenceManager().findPreference("reference_layer"));
        }
    }

    @Override
    public void onViewStateRestored(Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
        if (referenceLayerPref != null) {
            populateReferenceLayerPref();
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        referenceLayerPref = null;
    }

    /**
     * Creates the Basemap Source preference widget (but doesn't add it to
     * the screen; onBasemapSourceChanged will do that part).
     */
    private void initBasemapSourcePref() {
        basemapSourcePref = PrefUtils.createListPref(
            context, KEY_BASEMAP_SOURCE, getString(R.string.basemap_source),
            MapProvider.getLabelIds(), MapProvider.getIds()
        );
        basemapSourcePref.setIconSpaceReserved(false);
        onBasemapSourceChanged(MapProvider.getConfigurator());
        basemapSourcePref.setOnPreferenceChangeListener((pref, value) -> {
            MapConfigurator cftor = MapProvider.getConfigurator(value.toString());
            if (!cftor.isAvailable(context)) {
                cftor.showUnavailableMessage(context);
                return false;
            } else {
                onBasemapSourceChanged(cftor);
                return true;
            }
        });
    }

    /** Updates the rest of the preference UI when the Basemap Source is changed. */
    private void onBasemapSourceChanged(MapConfigurator cftor) {
        // Set up the preferences in the "Basemap" section.
        PreferenceCategory baseCategory = findPreference(CATEGORY_BASEMAP);
        baseCategory.removeAll();
        baseCategory.addPreference(basemapSourcePref);

        for (Preference pref : cftor.createPrefs(context)) {
            pref.setIconSpaceReserved(false);
            baseCategory.addPreference(pref);
        }

        // Clear the reference layer if it isn't supported by the new basemap.
        if (referenceLayerPref != null) {
            String path = referenceLayerPref.getValue();
            if (path != null && !cftor.supportsLayer(new File(path))) {
                referenceLayerPref.setValue(null);
                updateReferenceLayerSummary(null);
            }
        }
    }

    /** Sets up listeners for the Reference Layer preference widget. */
    private void initReferenceLayerPref() {
        referenceLayerPref = findPreference("reference_layer");
        referenceLayerPref.setOnPreferenceClickListener(preference -> {
            populateReferenceLayerPref();
            return false;
        });

        String layersPath = new StoragePathProvider().getOdkDirPath(StorageSubdirectory.LAYERS);

        if (referenceLayerPref.getValue() == null
                || new File(PathUtils.getAbsoluteFilePath(layersPath, referenceLayerPref.getValue())).exists()) {
            updateReferenceLayerSummary(referenceLayerPref.getValue());
        } else {
            referenceLayerPref.setValue(null);
            updateReferenceLayerSummary(null);
        }
        referenceLayerPref.setOnPreferenceChangeListener((preference, newValue) -> {
            updateReferenceLayerSummary(newValue);
            DialogFragment dialogFragment = (DialogFragment) getParentFragmentManager().findFragmentByTag(ReferenceLayerPreferenceDialog.class.getName());
            if (dialogFragment != null) {
                dialogFragment.dismiss();
            }
            return true;
        });
    }

    /** Sets the summary text for the reference layer to show the selected file. */
    private void updateReferenceLayerSummary(Object value) {
        if (referenceLayerPref != null) {
            String summary;
            if (value == null) {
                summary = getString(R.string.none);
            } else {
                MapConfigurator cftor = MapProvider.getConfigurator();
                String path = PathUtils.getAbsoluteFilePath(new StoragePathProvider().getOdkDirPath(StorageSubdirectory.LAYERS), value.toString());
                summary = cftor.getDisplayName(new File(path));
            }
            referenceLayerPref.setSummary(summary);
        }
    }


    /** Sets up the contents of the reference layer selection dialog. */
    private void populateReferenceLayerPref() {
        MapConfigurator cftor = MapProvider.getConfigurator();
        StoragePathProvider storagePathProvider = new StoragePathProvider();

        List<Item> items = new ArrayList<>();
        items.add(new Item(null, getString(R.string.none), ""));
        for (File file : getSupportedLayerFiles(cftor)) {
            String path = FileUtils.simplifyScopedStoragePath(file.getPath());
            String value = PathUtils.getRelativeFilePath(storagePathProvider.getOdkDirPath(StorageSubdirectory.LAYERS), file.getAbsolutePath());
            String name = cftor.getDisplayName(new File(file.getAbsolutePath()));
            items.add(new Item(value, name, path));
        }

        // Sort by display name, then by path for files with identical names.
        Collections.sort(items, (a, b) -> {
            if ((a.value == null) != (b.value == null)) {  // one or the other is null
                return a.value == null ? -1 : 1;
            }
            if (!a.label.equalsIgnoreCase(b.label)) {
                return a.label.compareToIgnoreCase(b.label);
            }
            if (!a.label.equals(b.label)) {
                return a.label.compareTo(b.label);
            }
            if (a.value != null && b.value != null) {
                return FileUtils.comparePaths(a.value, b.value);
            }
            return 0;  // both a.value and b.value are null
        });

        referenceLayerPref.setItems(items);

        String layerDir = FileUtils.simplifyScopedStoragePath(storagePathProvider.getOdkDirPath(StorageSubdirectory.LAYERS));
        referenceLayerPref.setDialogCaption(context.getString(
            items.size() > 1 ? R.string.layer_data_caption : R.string.layer_data_caption_none,
            layerDir, context.getString(MapProvider.getSourceLabelId())
        ));

        referenceLayerPref.updateContent();
    }

    /** Gets the list of layer data files supported by the current MapConfigurator. */
    private static List<File> getSupportedLayerFiles(MapConfigurator cftor) {
        List<File> files = new ArrayList<>();
        for (File file : FileUtils.walk(new File(new StoragePathProvider().getOdkDirPath(StorageSubdirectory.LAYERS)))) {
            if (cftor.supportsLayer(file)) {
                files.add(file);
            }
        }
        return files;
    }
}

File: collect_app/src/main/java/org/odk/collect/android/preferences/screens/MapsPreferencesFragment.kt
/*
 * Copyright (C) 2017 Shobhit
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.odk.collect.android.preferences.screens

import android.content.Context
import android.os.Bundle
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import org.odk.collect.android.R
import org.odk.collect.android.geo.MapConfiguratorProvider
import org.odk.collect.android.injection.DaggerUtils
import org.odk.collect.androidshared.ui.DialogFragmentUtils
import org.odk.collect.androidshared.ui.FragmentFactoryBuilder
import org.odk.collect.androidshared.ui.PrefUtils
import org.odk.collect.androidshared.ui.multiclicksafe.MultiClickGuard.allowClick
import org.odk.collect.async.Scheduler
import org.odk.collect.maps.MapConfigurator
import org.odk.collect.maps.layers.OfflineMapLayersPickerBottomSheetDialogFragment
import org.odk.collect.maps.layers.ReferenceLayerRepository
import org.odk.collect.settings.keys.ProjectKeys
import org.odk.collect.settings.keys.ProjectKeys.CATEGORY_BASEMAP
import org.odk.collect.settings.keys.ProjectKeys.KEY_BASEMAP_SOURCE
import org.odk.collect.strings.localization.getLocalizedString
import org.odk.collect.webpage.ExternalWebPageHelper
import javax.inject.Inject

class MapsPreferencesFragment : BaseProjectPreferencesFragment(), Preference.OnPreferenceClickListener {

    private lateinit var basemapSourcePref: ListPreference

    @Inject
    lateinit var referenceLayerRepository: ReferenceLayerRepository

    @Inject
    lateinit var scheduler: Scheduler

    @Inject
    lateinit var externalWebPageHelper: ExternalWebPageHelper

    override fun onAttach(context: Context) {
        super.onAttach(context)
        DaggerUtils.getComponent(context).inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        childFragmentManager.fragmentFactory = FragmentFactoryBuilder()
            .forClass(OfflineMapLayersPickerBottomSheetDialogFragment::class) {
                OfflineMapLayersPickerBottomSheetDialogFragment(requireActivity().activityResultRegistry, referenceLayerRepository, scheduler, settingsProvider, externalWebPageHelper)
            }
            .build()

        super.onCreate(savedInstanceState)
    }

    override fun onSettingChanged(key: String) {
        super.onSettingChanged(key)
        if (key == ProjectKeys.KEY_REFERENCE_LAYER) {
            findPreference<Preference>(ProjectKeys.KEY_REFERENCE_LAYER)!!.summary = getLayerName()
        } else if (key == KEY_BASEMAP_SOURCE) {
            val cftor = MapConfiguratorProvider.getConfigurator(settingsProvider.getUnprotectedSettings().getString(key))
            if (!cftor.isAvailable(requireContext())) {
                cftor.showUnavailableMessage(requireContext())
            } else {
                onBasemapSourceChanged(cftor)
            }
        }
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        super.onCreatePreferences(savedInstanceState, rootKey)
        setPreferencesFromResource(R.xml.maps_preferences, rootKey)
        initBasemapSourcePref()
        initLayersPref()
    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        if (allowClick(javaClass.name)) {
            when (preference.key) {
                ProjectKeys.KEY_REFERENCE_LAYER -> {
                    DialogFragmentUtils.showIfNotShowing(
                        OfflineMapLayersPickerBottomSheetDialogFragment::class.java,
                        childFragmentManager
                    )
                }
            }
            return true
        }
        return false
    }

    /**
     * Creates the Basemap Source preference widget (but doesn't add it to
     * the screen; onBasemapSourceChanged will do that part).
     */
    private fun initBasemapSourcePref() {
        basemapSourcePref = PrefUtils.createListPref(
            requireContext(),
            KEY_BASEMAP_SOURCE,
            getString(org.odk.collect.strings.R.string.basemap_source),
            MapConfiguratorProvider.getLabelIds(),
            MapConfiguratorProvider.getIds(),
            settingsProvider.getUnprotectedSettings()
        )

        basemapSourcePref.setIconSpaceReserved(false)
        onBasemapSourceChanged(MapConfiguratorProvider.getConfigurator())
    }

    private fun initLayersPref() {
        findPreference<Preference>(ProjectKeys.KEY_REFERENCE_LAYER)?.apply {
            onPreferenceClickListener = this@MapsPreferencesFragment
            summary = getLayerName()
        }
    }

    private fun getLayerName(): String {
        val layerId = settingsProvider.getUnprotectedSettings().getString(ProjectKeys.KEY_REFERENCE_LAYER)
        return if (layerId == null) {
            requireContext().getLocalizedString(org.odk.collect.strings.R.string.none)
        } else {
            referenceLayerRepository.get(layerId)!!.name
        }
    }

    /** Updates the rest of the preference UI when the Basemap Source is changed.  */
    private fun onBasemapSourceChanged(cftor: MapConfigurator) {
        // Set up the preferences in the "Basemap" section.
        val baseCategory = findPreference<PreferenceCategory>(CATEGORY_BASEMAP)
        baseCategory!!.removeAll()
        baseCategory.addPreference(basemapSourcePref)
        for (pref in cftor.createPrefs(requireContext(), settingsProvider.getUnprotectedSettings())) {
            pref.isIconSpaceReserved = false
            baseCategory.addPreference(pref)
        }

        // Clear the reference layer if it does not exist or it isn't supported by the new basemap.
        val layerId = settingsProvider.getUnprotectedSettings().getString(ProjectKeys.KEY_REFERENCE_LAYER)
        if (layerId != null) {
            val layer = referenceLayerRepository.get(layerId)
            if (layer == null) {
                settingsProvider.getUnprotectedSettings().save(ProjectKeys.KEY_REFERENCE_LAYER, null)
            }
        }
    }
}
